### PR TITLE
[Codegen] Extract the parseFile function in the typescript and flow parsers

### PR DESCRIPTION
**Pull request by MaeIg**
## Summary

This PR aims to extract  the parseFile function in the typescript and flow parsers.  This is to solve the problem described [here](https://github.com/facebook/react-native/pull/35158#issuecomment-1298330753) and help with the work done in #34872.

## Changelog

[Internal] [Changed] - Extract the parseFile function in the typescript and flow parsers

## Test Plan

yarn flow:
image

yarn lint:
image

yarn test:
image

### CODE REVIEW

**Review by cipolleschi — APPROVED** [review]
There are many changes on several files, but it looks very good to me. I like that we are centralizing the shared logic in a specific object.

Thank you so much for taking care of this.

**Comment by cipolleschi** [comment]
There is a problem with the BUCK tests, though... :/

**Comment by MaeIg** [comment]
I think there is a circular dependency problem :/ 
image

**Comment by MaeIg** [comment]
> There is a problem with the BUCK tests, though... :/

@cipolleschi, I think the errors come from circular imports.
1. In FlowParser, we import buildSchema from packages/react-native-codegen/src/parsers/flow/index.js
2. In this file, we import buildModuleSchema from packages/react-native-codegen/src/parsers/flow/modules/index.js
3. In this file, we import FlowParser

It will be solved in #35147. But, I think the problem can come back later and I don't know how to fix it.

Do you know how we could avoid these circular imports? Aren't we importing the parser in too many places?

**Comment by Pranav-yadav** [comment]
@MaeIg I'm also getting the same warnings w/ circular dependencies here https://github.com/facebook/react-native/pull/35288#issuecomment-1312414197

I think the dependents should be moved appropriately to resolve this.. Though there are already 2-3 PRs which are dependent.. it would be fixed if one of them gets merged. 👍

**Comment by cipolleschi** [comment]
Hi @MaeIg, could you rebase this, whenever you have some time, please? 🙏

**Comment by MaeIg** [comment]
> Hi @MaeIg, could you rebase this, whenever you have some time, please? 🙏

Hi, sorry I was pretty busy this week.  I will do it tomorrow!

**Comment by MaeIg** [comment]
I rebased but there is still a circular dependency:
image
(there is the same for typescript)

It seems complicated to undo these circular dependencies. I wonder if it's time to extract parseFile in the parser. Maybe I should do the other task without it.

**Comment by MaeIg** [comment]
> What if we don't `require` the `FlowParser` in the `flow/module/index.js` file, but we just pass it as a parameter to the methods that need it? 🤔 We can just `import type`, if we want to be stricter with the types, or just ask for a `Parser`. At this point, the `parsers/index.js` module can create the `FlowParser` and pass it to the methods but we break the `require` depenencies. What do you think?

I will try to do it. It will make a lot of changes but this rule should avoid circular dependencies!

**Comment by Pranav-yadav** [comment]
@MaeIg, I've suggested following; under PR which resolved Circular Dependencies (similar issues); which @cipolleschi might've missed, here: https://github.com/facebook/react-native/pull/35288#issuecomment-1320051562:-

> We should probably define some clear rules about what should go in parsers-primitives and what should go in parsers-commons... 🤔 Probably, it's better to put in primitives all the basic building blocks and commons all the utility functions, so that primitives can import commons but vice-versa should not happen.
What do you think? --- _Riccardo_
> > Definitely. Block comments should be added at top of both parsers-commons.js and parsers-primitives.js stating their purpose, functionalities, and what NOT to include in those files.. :) Also, It would be very useful for newbie contributors like me 👶 as well as existing.. Thanks --- _Pranav_
> > > @cipolleschi After giving it another thought, I think we should merge those two files. Reason: Cohesion
> > > Though, (logically) it's correct to keep them separate, the fns/types/interfaces exported by them are not used sparsely (at different places).
We know that, they are highly related to each other (highly cohesive => Good 🙂) and keeping logic together will be better as well as easy to maintain.
Thoughts?
> > > PS: sorry for commenting under closed PR -- _Pranav_

**Comment by cipolleschi** [comment]
@Pranav-yadav, sorry if I reply here and I haven't replied in the other PR. I understand what you say and it makes a lot of sense. I'm quite scared about how big the file will ends up. It could become very hard to read an navigate. 🤔 That's the main reason why I was pushing to keep the files separated.

However, we currently have a different circular dependency: it is not related to `parsers-commons` and `-primitives` but to the Parser itself and the `index` files.

**Comment by MaeIg** [comment]
Hi @cipolleschi,

After a hard fight I ended up overcoming these circular dependencies! 🎉 

I recommend you to review commit by commit to understand well the refacto 😃 

Here's a diagram I created to understand what to do. Maybe it can help you to review!
rn-circular-deps

File: packages/eslint-plugin-specs/react-native-modules.js [hunk 1]
```diff
@@ -36,6 +37,7 @@ function requireModuleParser() {
       withBabelRegister(config, () => {
         RNModuleParser = require('@react-native/codegen/src/parsers/flow/modules');
         RNParserUtils = require('@react-native/codegen/src/parsers/utils');
+        RNFlowParser = require('@react-native/codegen//src/parsers/flow/parser');
```
Comment (cipolleschi) [inline packages/eslint-plugin-specs/react-native-modules.js:40]:
```suggestion
        RNFlowParser = require('@react-native/codegen/src/parsers/flow/parser');
```

File: packages/react-native-codegen/src/parsers/typescript/buildSchema.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    !contents.includes('codegenNativeComponent') &&
+    !contents.includes('TurboModule')
+  ) {
+    return {modules: {}};
+  }
+
+  const ast = babelParser.parse(contents, {
+    sourceType: 'module',
+    plugins: ['typescript'],
+  }).program;
```
Comment (cipolleschi) [inline packages/react-native-codegen/src/parsers/typescript/buildSchema.js:71]: What about creating a function in the `Flow/TypeScriptParser` called `getAST()` that uses the `flow-parser` or the `babelParser` and return an untyped `ast`? 🤔 
If I'm not reading this wrong, that's the only difference between the two `buildSchema` functions. With this change, we should be able to have a common `buildSchema` function that asks to the parser to get the proper AST.

How do you feel about this change?

Comment (cipolleschi) [reply packages/react-native-codegen/src/parsers/typescript/buildSchema.js:71]: If we go for this path, we can rename the `buildSchema.js` to `Visitor.js` and the two files will contain just the Flow/TypeScript implementation for the visitors. :D

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @MaeIg, thank you endlessly for taking care of this and for all these changes. 🙏 

I left some comments to improve this even further, but it already looks extremely good to me.

File: packages/eslint-plugin-specs/react-native-modules.js (hunk 1 above)
Comment (MaeIg) [reply packages/eslint-plugin-specs/react-native-modules.js:40]: Ups my bad, I rebased too quickly. Thanks for the catch!

File: packages/react-native-codegen/src/parsers/typescript/buildSchema.js (hunk 2 above)
Comment (cipolleschi) [reply packages/react-native-codegen/src/parsers/typescript/buildSchema.js:71]: that's ok, then!

**Comment by cipolleschi** [comment]
This PR needs some changes internally. Unfortunately, I don't think I'll be able to to them today, so we will merge it in a week. I'm sorry for this. :(

**Comment by MaeIg** [comment]
Hello 👋
i rebased and resolved conflicts. Don't hesitate to ping me if there is anything I can do

**Review by cipolleschi — APPROVED** [review]

**Pull request closed without merging**